cmd/kubeadm/app/phases/addons/dns/dns.go

```diff
@@ -80,15 +79,36 @@ func DeployedDNSAddon(client clientset.Interface) (kubeadmapi.DNSAddOnType, stri
 	}
 }
 
+// deployedDNSReplicas returns the replica count for the current DNS deployment
+func deployedDNSReplicas(client clientset.Interface, replicas int32) (*int32, error) {
+	deploymentsClient := client.AppsV1().Deployments(metav1.NamespaceSystem)
+	deployments, err := deploymentsClient.List(metav1.ListOptions{LabelSelector: "k8s-app=kube-dns"})
+	if err != nil {
+		return &replicas, errors.Wrap(err, "couldn't retrieve DNS addon deployments")
+	}
+	switch len(deployments.Items) {
+	case 0:
+		return &replicas, errors.Wrap(err, "no DNS addons found")
+	case 1:
+		return deployments.Items[0].Spec.Replicas, nil
+	default:
+		return &replicas, errors.Errorf("multiple DNS addon deployments found: %v", deployments.Items)
+	}
+}
+
 // EnsureDNSAddon creates the kube-dns or CoreDNS addon
 func EnsureDNSAddon(cfg *kubeadmapi.ClusterConfiguration, client clientset.Interface) error {
+	replicas, err := deployedDNSReplicas(client, kubeadmconstants.CoreDNSReplicas)
+	if err != nil {
+		return err
+	}
 	if cfg.DNS.Type == kubeadmapi.CoreDNS {
-		return coreDNSAddon(cfg, client)
+		return coreDNSAddon(cfg, client, replicas)
 	}
-	return kubeDNSAddon(cfg, client)
+	return kubeDNSAddon(cfg, client, replicas)
```

Comment:
This is going to result in a couple of kube-dns replicas (instead of 1 as it is today).

Reply:
I have updated this. But don't like that we are calling two funcs that ultimately could be merged into 1. Waiting to hear back on the other comment for the best way forward on this.

Reply:
I have shuffled this around to use each of the DNS options defaults. I will wait to see your thoughts on calling two funcs that ultimately do the same thing and the result of your other open comment above and we can go from there in making this nicer.